sdks/java/extensions/sorter/src/main/java/org/apache/beam/sdk/extensions/sorter/InMemorySorter.java

```diff
@@ -0,0 +1,169 @@
+
+  /** Private constructor. */
+  private InMemorySorter(Options options) {
+    maxBufferSize = options.getMemoryMB() * 1024 * 1024;
+  }
+
+  /** Create a new sorter from provided options. */
+  public static InMemorySorter create(Options options) {
+    return new InMemorySorter(options);
+  }
+
+  @Override
+  public void add(KV<byte[], byte[]> record) {
+    boolean successful = addIfRoom(record);
+    if (!successful) {
+      throw new IllegalStateException("No space remaining for in memory sorting");
+    }
+  }
+
+  /** Adds the record is there is room and returns true. Otherwise returns false. */
+  public boolean addIfRoom(KV<byte[], byte[]> record) {
+    if (sortCalled) {
+      throw new IllegalStateException("Records can only be added before sort().");
+    }
+
+    int recordBytes = estimateRecordBytes(record);
+    if (roomInBuffer(numBytes + recordBytes, records.size() + 1)) {
+      records.add(record);
+      numBytes += recordBytes;
+      return true;
+    } else {
+      return false;
+    }
+  }
+
+  @Override
+  public Iterable<KV<byte[], byte[]>> sort() {
+    if (sortCalled) {
+      throw new IllegalStateException("sort() can only be called once.");
+    }
+
+    sortCalled = true;
+
+    Comparator<KV<byte[], byte[]>> kvComparator =
+        new Comparator<KV<byte[], byte[]>>() {
+
+          @Override
+          public int compare(KV<byte[], byte[]> o1, KV<byte[], byte[]> o2) {
+            return COMPARATOR.compare(o1.getKey(), o2.getKey());
+          }
+        };
+    Collections.sort(records, kvComparator);
+    return Collections.unmodifiableList(records);
+  }
+
+  /**
+   * Estimate the number of additional bytes required to store this record. Including the key, the
+   * value and any overhead for objects and references.
+   */
+  private int estimateRecordBytes(KV<byte[], byte[]> record) {
+    return RECORD_MEMORY_OVERHEAD_ESTIMATE + record.getKey().length + record.getValue().length;
+  }
+
+  /**
+   * Check whether we have room to store the provided total number of bytes and total number of
+   * records.
+   */
+  private boolean roomInBuffer(int numBytes, int numRecords) {
+    // Collections.sort may allocate up to n/2 extra object references.
+    // Also, ArrayList grows by a factor of 1.5x, so there might be up to n/2 null object
+    // references in the backing array.
+    // And finally, in Java 7, Collections.sort performs a defensive copy to an array in case the
+    // input list is a LinkedList.
+    // So assume we need an additional overhead of two words per record in the worst case
+    return (numBytes + (numRecords * NUM_BYTES_PER_WORD * 2)) < maxBufferSize;
+  }
+
+  /**
+   * Returns the number of bytes in a word according to the JVM. Defaults to 8 for 64 bit if answer
+   * unknown.
+   */
+  private static int getNumBytesPerWord() {
+    String bitsPerWord = System.getProperty("sun.arch.data.model");
```

Comment:
I checked that myself before submitting the PR and parseInt does actually throw a NumberFormatException if the argument is null. However, that isn't explicitly documented and even if it were, it could conceivably change in a future version of java. So took your advice below of guarding against Exception.